File: packages/runtime-web-three/src/spawner.ts
import type { IPrefabsIr, ISpawnerComponent, ITransformComponent, IWorldEntity, IWorldIr, Vec3 } from "@threenative/ir";

interface ISpawnerState {
  nextTick: number;
  sequence: number;
  total: number;
}

export interface ISpawnerObservation {
  entity: string;
  prefab: string;
  root: string;
  spawned: readonly string[];
  tick: number;
}

const spawnerStateByWorld = new WeakMap<IWorldIr, Map<string, ISpawnerState>>();

export function hasSpawners(world: IWorldIr): boolean {
  return world.entities.some((entity) => entity.components.Spawner !== undefined);
}

export function stepSpawners(world: IWorldIr, options: { fixedDelta?: number; prefabs?: IPrefabsIr; tick: number }): ISpawnerObservation[] {
  const state = spawnerStateByWorld.get(world) ?? new Map<string, ISpawnerState>();
  spawnerStateByWorld.set(world, state);
  const observations: ISpawnerObservation[] = [];
  const fixedDelta = options.fixedDelta ?? 1 / 60;

  for (const entity of [...world.entities]) {
    const spawner = entity.components.Spawner;
    if (spawner === undefined || spawner.enabled === false) {
      continue;
    }
    const template = options.prefabs?.prefabs.find((candidate) => candidate.id === spawner.prefab);
    if (template === undefined) {
      continue;
    }
    const spawnerState = stateFor(entity.id, spawner, fixedDelta, state);
    if (options.tick < spawnerState.nextTick || isDepleted(spawner, spawnerState)) {
      continue;
    }
    const alive = aliveSpawnCount(world, entity.id);
    const capacity = Math.max(0, (spawner.maxAlive ?? Number.POSITIVE_INFINITY) - alive);
    const batchSize = Math.min(capacity, batchSizeFor(spawner), remainingFor(spawner, spawnerState));
    if (batchSize <= 0) {
      continue;
    }
    for (let index = 0; index < batchSize; index += 1) {
      const sequence = spawnerState.sequence;
      const prefix = `${entity.id}.spawn.${sequence}`;
      const spawned = instantiatePrefab(world, template, prefix, entity, spawner, sequence);
      if (spawned.length > 0) {
        observations.push({ entity: entity.id, prefab: spawner.prefab, root: `${prefix}.${template.root}`, spawned, tick: options.tick });
        spawnerState.total += 1;
        spawnerState.sequence += 1;
      }
    }
    spawnerState.nextTick = nextTickFor(spawner, fixedDelta, options.tick);
  }

  if (observations.length > 0) {
    appendSpawnerEvents(world, "spawner.spawned", observations.map((observation) => ({ entity: observation.entity, prefab: observation.prefab, root: observation.root, tick: observation.tick })));
  }
  return observations;
}

function stateFor(id: string, spawner: ISpawnerComponent, fixedDelta: number, states: Map<string, ISpawnerState>): ISpawnerState {
  const existing = states.get(id);
  if (existing !== undefined) {
    return existing;
  }
  const initial = { nextTick: 0, sequence: 0, total: 0 };
  if (spawner.mode === "interval" && spawner.interval !== undefined) {
    initial.nextTick = intervalTicks(spawner.interval, fixedDelta);
  }
  states.set(id, initial);
  return initial;
}

function batchSizeFor(spawner: ISpawnerComponent): number {
  if (spawner.mode === "interval") {
    return 1;
  }
  return Math.max(1, Math.floor(spawner.waveSize ?? 1));
}

function remainingFor(spawner: ISpawnerComponent, state: ISpawnerState): number {
  return Math.max(0, (spawner.maxTotal ?? Number.POSITIVE_INFINITY) - state.total);
}

function isDepleted(spawner: ISpawnerComponent, state: ISpawnerState): boolean {
  return spawner.maxTotal !== undefined && state.total >= spawner.maxTotal;
}

function nextTickFor(spawner: ISpawnerComponent, fixedDelta: number, currentTick: number): number {
  if (spawner.mode === "once" || (spawner.mode === "wave" && spawner.interval === undefined)) {
    return Number.POSITIVE_INFINITY;
  }
  return currentTick + intervalTicks(spawner.interval ?? fixedDelta, fixedDelta);
}

function intervalTicks(interval: number, fixedDelta: number): number {
  return Math.max(1, Math.ceil(interval / Math.max(fixedDelta, 1 / 600)));
}

function aliveSpawnCount(world: IWorldIr, spawnerId: string): number {
  const prefix = `${spawnerId}.spawn.`;
  return world.entities.filter((entity) => entity.id.startsWith(prefix)).length;
}

function instantiatePrefab(world: IWorldIr, prefab: IPrefabsIr["prefabs"][number], prefix: string, spawner: IWorldEntity, config: ISpawnerComponent, sequence: number): string[] {
  const spawned: string[] = [];
  const offset = spawnOffset(config, sequence);
  const origin = vector(spawner.components.Transform?.position);
  for (const template of prefab.entities) {
    const id = `${prefix}.${template.id}`;
    if (world.entities.some((entity) => entity.id === id)) {
      continue;
    }
    const components = clone(template.components) as IWorldEntity["components"];
    components.Transform = offsetTransform(components.Transform, origin, offset);
    world.entities.push({ components, id });
    spawned.push(id);
  }
  return spawned;
}

function offsetTransform(transform: ITransformComponent | undefined, origin: Vec3, offset: Vec3): ITransformComponent {
  const position = vector(transform?.position);
  return {
    ...(transform ?? {}),
    position: [position[0] + origin[0] + offset[0], position[1] + origin[1] + offset[1], position[2] + origin[2] + offset[2]],
  };
}

function spawnOffset(spawner: ISpawnerComponent, sequence: number): Vec3 {
  const area = spawner.area;
  if (area === undefined || area.shape === "point") {
    return [0, 0, 0];
  }
  const rng = seededRandom((spawner.jitterSeed ?? 0) + sequence * 1013);
  if (area.shape === "circle") {
    const radius = typeof area.size === "number" ? area.size : vector2(area.size)[0];
    const angle = rng() * Math.PI * 2;
    const distance = Math.sqrt(rng()) * radius;
    return [Math.cos(angle) * distance, 0, Math.sin(angle) * distance];
  }
  const size = vector(area.size);
  return [(rng() - 0.5) * size[0], (rng() - 0.5) * size[1], (rng() - 0.5) * size[2]];
}

function appendSpawnerEvents(world: IWorldIr, event: string, payloads: readonly unknown[]): void {
  const queue = world.events?.[event];
  world.events = {
    ...(world.events ?? {}),
    [event]: Array.isArray(queue) ? [...queue, ...payloads] : [...payloads],
  };
}

function seededRandom(seed: number): () => number {
  let state = (Math.floor(seed) >>> 0) || 1;
  return () => {
    state = (1664525 * state + 1013904223) >>> 0;
    return state / 0x100000000;
  };
}

function vector(value: unknown): Vec3 {
  return Array.isArray(value) && value.length === 3 && value.every((item) => typeof item === "number" && Number.isFinite(item))
    ? [value[0], value[1], value[2]]
    : [0, 0, 0];
}

function vector2(value: unknown): readonly [number, number] {
  return Array.isArray(value) && value.length === 2 && value.every((item) => typeof item === "number" && Number.isFinite(item))
    ? [value[0], value[1]]
    : [0, 0];
}

function clone<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

File: runtime-bevy/crates/threenative_runtime/src/spawner.rs
use std::collections::BTreeMap;

use serde::Serialize;
use serde_json::json;
use threenative_loader::{
    EntityComponents, PrefabDeclarationIr, PrefabsIr, SpawnerComponent, TransformComponent,
    WorldEntity, WorldIr,
};

#[derive(Clone, Debug, Default)]
pub struct NativeSpawnerRuntimeState {
    spawners: BTreeMap<String, NativeSpawnerState>,
}

#[derive(Clone, Debug)]
struct NativeSpawnerState {
    next_tick: u32,
    sequence: u32,
    total: u32,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct NativeSpawnerObservation {
    pub entity: String,
    pub prefab: String,
    pub root: String,
    pub spawned: Vec<String>,
    pub tick: u32,
}

pub fn step_world_spawners(
    world: &mut WorldIr,
    prefabs: Option<&PrefabsIr>,
    tick: u32,
    fixed_delta: f32,
    state: &mut NativeSpawnerRuntimeState,
) -> Vec<NativeSpawnerObservation> {
    let mut observations = Vec::new();
    let source_entities = world.entities.clone();
    for entity in source_entities {
        let Some(spawner) = entity.components.spawner.clone() else {
            continue;
        };
        if !spawner.enabled {
            continue;
        }
        let Some(prefab) = prefabs.and_then(|items| {
            items
                .prefabs
                .iter()
                .find(|candidate| candidate.id == spawner.prefab)
        }) else {
            continue;
        };
        let spawner_state = state_for(&entity.id, &spawner, fixed_delta, state);
        if tick < spawner_state.next_tick || is_depleted(&spawner, spawner_state) {
            continue;
        }
        let capacity = spawner
            .max_alive
            .unwrap_or(u32::MAX)
            .saturating_sub(alive_spawn_count(world, &entity.id));
        let batch_size = capacity
            .min(batch_size_for(&spawner))
            .min(remaining_for(&spawner, spawner_state));
        if batch_size == 0 {
            continue;
        }
        for _ in 0..batch_size {
            let sequence = spawner_state.sequence;
            let prefix = format!("{}.spawn.{}", entity.id, sequence);
            let spawned = instantiate_prefab(world, prefab, &prefix, &entity, &spawner, sequence);
            if !spawned.is_empty() {
                observations.push(NativeSpawnerObservation {
                    entity: entity.id.clone(),
                    prefab: spawner.prefab.clone(),
                    root: format!("{}.{}", prefix, prefab.root),
                    spawned,
                    tick,
                });
                spawner_state.total += 1;
                spawner_state.sequence += 1;
            }
        }
        spawner_state.next_tick = next_tick_for(&spawner, fixed_delta, tick);
    }

    if !observations.is_empty() {
        let payload = observations
            .iter()
            .map(|observation| {
                json!({
                    "entity": observation.entity,
                    "prefab": observation.prefab,
                    "root": observation.root,
                    "tick": observation.tick,
                })
            })
            .collect::<Vec<_>>();
        world
            .events
            .entry("spawner.spawned".to_owned())
            .and_modify(|value| {
                if let Some(items) = value.as_array_mut() {
                    items.extend(payload.clone());
                }
            })
            .or_insert_with(|| serde_json::Value::Array(payload));
    }
    observations
}

fn state_for<'a>(
    id: &str,
    spawner: &SpawnerComponent,
    fixed_delta: f32,
    runtime: &'a mut NativeSpawnerRuntimeState,
) -> &'a mut NativeSpawnerState {
    runtime
        .spawners
        .entry(id.to_owned())
        .or_insert_with(|| NativeSpawnerState {
            next_tick: if spawner.mode == "interval" {
                interval_ticks(spawner.interval.unwrap_or(fixed_delta), fixed_delta)
            } else {
                0
            },
            sequence: 0,
            total: 0,
        })
}

fn batch_size_for(spawner: &SpawnerComponent) -> u32 {
    if spawner.mode == "interval" {
        1
    } else {
        spawner.wave_size.unwrap_or(1).max(1)
    }
}

fn remaining_for(spawner: &SpawnerComponent, state: &NativeSpawnerState) -> u32 {
    spawner
        .max_total
        .unwrap_or(u32::MAX)
        .saturating_sub(state.total)
}

fn is_depleted(spawner: &SpawnerComponent, state: &NativeSpawnerState) -> bool {
    spawner
        .max_total
        .is_some_and(|max_total| state.total >= max_total)
}

fn next_tick_for(spawner: &SpawnerComponent, fixed_delta: f32, current_tick: u32) -> u32 {
    if spawner.mode == "once" || (spawner.mode == "wave" && spawner.interval.is_none()) {
        u32::MAX
    } else {
        current_tick.saturating_add(interval_ticks(
            spawner.interval.unwrap_or(fixed_delta),
            fixed_delta,
        ))
    }
}

fn interval_ticks(interval: f32, fixed_delta: f32) -> u32 {
    (interval / fixed_delta.max(1.0 / 600.0)).ceil().max(1.0) as u32
}

fn alive_spawn_count(world: &WorldIr, spawner_id: &str) -> u32 {
    let prefix = format!("{}.spawn.", spawner_id);
    world
        .entities
        .iter()
        .filter(|entity| entity.id.starts_with(&prefix))
        .count() as u32
}

fn instantiate_prefab(
    world: &mut WorldIr,
    prefab: &PrefabDeclarationIr,
    prefix: &str,
    spawner: &WorldEntity,
    config: &SpawnerComponent,
    sequence: u32,
) -> Vec<String> {
    let mut spawned = Vec::new();
    let offset = spawn_offset(config, sequence);
    let origin = spawner
        .components
        .transform
        .as_ref()
        .and_then(|transform| transform.position)
        .unwrap_or([0.0, 0.0, 0.0]);
    for template in &prefab.entities {
        let id = format!("{}.{}", prefix, template.id);
        if world.entities.iter().any(|entity| entity.id == id) {
            continue;
        }
        let mut components = template.components.clone();
        components.transform = Some(offset_transform(
            components.transform.as_ref(),
            origin,
            offset,
        ));
        world.entities.push(WorldEntity {
            id: id.clone(),
            components,
            tags: Vec::new(),
        });
        spawned.push(id);
    }
    spawned
}

fn offset_transform(
    transform: Option<&TransformComponent>,
    origin: [f32; 3],
    offset: [f32; 3],
) -> TransformComponent {
    let position = transform
        .and_then(|value| value.position)
        .unwrap_or([0.0, 0.0, 0.0]);
    TransformComponent {
        position: Some([
            position[0] + origin[0] + offset[0],
            position[1] + origin[1] + offset[1],
            position[2] + origin[2] + offset[2],
        ]),
        rotation: transform.and_then(|value| value.rotation),
        scale: transform.and_then(|value| value.scale),
    }
}

fn spawn_offset(spawner: &SpawnerComponent, sequence: u32) -> [f32; 3] {
    let Some(area) = &spawner.area else {
        return [0.0, 0.0, 0.0];
    };
    if area.shape == "point" {
        return [0.0, 0.0, 0.0];
    }
    let mut random = SeededRandom::new(
        spawner.jitter_seed.unwrap_or(0.0) as u32 + sequence.saturating_mul(1013),
    );
    if area.shape == "circle" {
        let radius = area_size_number(area.size.as_ref());
        let angle = random.next() * std::f32::consts::TAU;
        let distance = random.next().sqrt() * radius;
        return [angle.cos() * distance, 0.0, angle.sin() * distance];
    }
    let size = area_size_vec3(area.size.as_ref());
    [
        (random.next() - 0.5) * size[0],
        (random.next() - 0.5) * size[1],
        (random.next() - 0.5) * size[2],
    ]
}

fn area_size_number(value: Option<&serde_json::Value>) -> f32 {
    value
        .and_then(|item| item.as_f64())
        .map(|item| item as f32)
        .or_else(|| {
            value
                .and_then(|item| item.as_array())
                .and_then(|items| items.first())
                .and_then(|item| item.as_f64())
                .map(|item| item as f32)
        })
        .unwrap_or(0.0)
}

fn area_size_vec3(value: Option<&serde_json::Value>) -> [f32; 3] {
    let Some(items) = value.and_then(|item| item.as_array()) else {
        return [0.0, 0.0, 0.0];
    };
    [
        items.first().and_then(|item| item.as_f64()).unwrap_or(0.0) as f32,
        items.get(1).and_then(|item| item.as_f64()).unwrap_or(0.0) as f32,
        items.get(2).and_then(|item| item.as_f64()).unwrap_or(0.0) as f32,
    ]
}

struct SeededRandom {
    state: u32,
}

impl SeededRandom {
    fn new(seed: u32) -> Self {
        Self {
            state: if seed == 0 { 1 } else { seed },
        }
    }

    fn next(&mut self) -> f32 {
        self.state = self
            .state
            .wrapping_mul(1_664_525)
            .wrapping_add(1_013_904_223);
        self.state as f32 / 4_294_967_296.0
    }
}

#[allow(dead_code)]
fn _assert_components_clone(_: EntityComponents) {}
